tensorflow/compiler/tf2tensorrt/convert/convert_nodes.cc

```diff
@@ -1535,6 +1535,71 @@ class StaticAlgorithmSelector : public nvinfer1::IAlgorithmSelector {
                         const nvinfer1::IAlgorithm* const* algoChoices,
                         int32_t nbAlgorithms) override {}  // do nothing
 };
+
+enum class Tactic : int64_t { kINVALID_TACTIC = int64_t(0xD15EA5EDD15EA5ED) };
+
+enum class LayerImpl : int64_t {
+  kSHUFFLE = 16,
+};
+
+static std::unordered_set<int64> trt7_bad_tactics{
+    -5927686925093575778, -3848538574386518527, -959009792490796596};
+
+class FlakyAlgorithmRejector : public nvinfer1::IAlgorithmSelector {
+ public:
+  FlakyAlgorithmRejector() = default;
+  ~FlakyAlgorithmRejector() = default;
+
+  // Returns value in [0, nbChoices] for a valid algorithm.
+  int32_t selectAlgorithms(const nvinfer1::IAlgorithmContext& algoContext,
+                           const nvinfer1::IAlgorithm* const* algoChoices,
+                           int32_t nbChoices, int32_t* selection) override {
+    bool reject{false};
+    // TensorRT always provides more than zero number of algorithms
+    // in selectAlgorithms.
+    assert(nbChoices > 0);
+
+    // Disable bad known tactics
+    int32_t nbSelections = 0;
+    for (auto i = 0; i < nbChoices; i++) {
+      if (algoChoices[i] == nullptr) {
+        continue;
+      }
+
+      int64 tacticId = algoChoices[i]->getAlgorithmVariant().getTactic();
+      int64 implementation =
+          algoChoices[i]->getAlgorithmVariant().getImplementation();
+      nvinfer1::TensorFormat format =
+          algoChoices[i]->getAlgorithmIOInfo(0).getTensorFormat();
+      assert(tacticId != static_cast<int64_t>(Tactic::kINVALID_TACTIC));
+
+#if !IS_TRT_VERSION_GE(8, 0, 0, 0)
+      if (trt7_bad_tactics.find(tacticId) != trt7_bad_tactics.end()) {
+        reject = true;
+      }
+      if (implementation == static_cast<int64_t>(LayerImpl::kSHUFFLE) &&
+          format == nvinfer1::TensorFormat::kCHW32) {
+        reject = true;
+      }
+#endif
+
+      if (reject) {
+        VLOG(1) << "Rejecting a bad known tactic: " << tacticId << " for node "
```

Comment:
let's remove "a bad known". This feature can be used for debugging, not necessarily because some tactic is known bad

Reply:
Interesting comment, but things here are already hard coded for the known bad tactics in TRT7.
I think we should first try to make the implementation more flexible and then fix the log message to reflect the situation.

Reply:
Good point @bixia1. Please ignore my comment

Reply:
Addressed this issue by replacing "a bad known" with "a disallowed tactic".